### PR TITLE
Adding `qml.Superposition`

### PR DESCRIPTION
**Pull request by KetpuntoG**
**Context:**

This PR implements `qml.Superposition,` a new operator that prepares states  given as superposition of basic states

[sc-72648]

**Benefits:**

Allow user to prepare states in a new format
In addition, the decomposition is efficient for this type of states

**Possible Drawbacks:**

### CODE REVIEW

File: pennylane/templates/state_preparations/superposition.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+    binary_dict = {}
+    used_smallest = set()
+
+    # Assign keys that can map to themselves
+    for original in basis_list:
+
+        if original in smallest_basis_lists and tuple(original) not in used_smallest:
+
+            binary_dict[tuple(original)] = original
+            used_smallest.add(tuple(original))
```
Comment (willjmax) [inline pennylane/templates/state_preparations/superposition.py:67]: So, if a state appears in `basis_list` multiple times the duplicates are just ignored? Maybe this should raise an exception because it doesn't make sense to put two copies of the same state in superposition.

File: pennylane/templates/state_preparations/superposition.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+            value
+            for key, value in sorted(
+                new_dic_state.items(), key=lambda item: int("".join(map(str, item[0])), 2)
+            )
+        ]
+
+        op_list = []
+        op_list.append(
+            qml.StatePrep(
+                qml.math.stack(sorted_coefficients),
```
Comment (willjmax) [inline pennylane/templates/state_preparations/superposition.py:264]: Why do we need this? Is passing in the list not sufficient?

File: pennylane/templates/state_preparations/superposition.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+            CNOT(wires=[2, 0]),
+            Toffoli(wires=[0, 1, 2])]
+
+        """
+
+        dic_state = dict(zip(basis, coefs))
+        perms = _get_permutation(basis)
+        new_dic_state = {perms[key]: dic_state[key] for key in dic_state if key in perms}
+
+        sorted_coefficients = [
```
Comment (willjmax) [inline pennylane/templates/state_preparations/superposition.py:254]: Should we check the norm of the coefficients? or will that just happen in `StatePrep`?

**Review by willjmax — APPROVED** [review]
Looks good. Just some minor comments.

File: pennylane/templates/state_preparations/superposition.py (hunk 1 above)
Comment (KetpuntoG) [reply pennylane/templates/state_preparations/superposition.py:67]: That's a good idea, I'll add a error if the basis provided is not unique!

File: pennylane/templates/state_preparations/superposition.py (hunk 3 above)
Comment (KetpuntoG) [reply pennylane/templates/state_preparations/superposition.py:254]: We assume that they are normalized. If not, StatePrep will display an error

File: pennylane/templates/state_preparations/superposition.py (hunk 2 above)
Comment (KetpuntoG) [reply pennylane/templates/state_preparations/superposition.py:264]: An appeared when working with ArrayBoxes because the input was a list of them, instead of an ArrayBox with a list inside. This line fixes it

File: pennylane/templates/state_preparations/superposition.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+# limitations under the License.
+r"""
+Contains the Superposition template.
+"""
+
+import pennylane as qml
+from pennylane.operation import AnyWires, Operation
+
+
+def _get_permutation(basis_list):
```
[the remark this answers is not in the document]
Comment (KetpuntoG) [reply pennylane/templates/state_preparations/superposition.py:22]: changed to `_assign_states`

File: pennylane/templates/state_preparations/superposition.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+            return qml.probs(wires)
+
+    .. code-block:: pycon
+
+        >>> print(circuit())
+        [0.33333333 0.         0.33333333 0.         0.         0.
+        0.         0.33333333]
+
+
+    .. details::
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:156]: Please add a bit more details to this section. Assume that the user doesn't know much about state preparation.

File: pennylane/templates/state_preparations/superposition.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+
+    @staticmethod
+    def compute_decomposition(coefs, basis, wires, work_wire):  # pylint: disable=arguments-differ
+        r"""Representation of the operator as a product of other operators.
+
+        Args:
+            coefs (List[float]): List of coefficients :math:`c_i`
+            basis (List[List[int]]): List of basis states :math:`|b_i\rangle`
+            wires (Sequence[int]): List of wires that the operator acts on
+            work_wire (Union[Wires, int, str]): The auxiliary wire used for the permutation
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:240]:
```suggestion
            coefs (List[float]): list of coefficients :math:`c_i`
            basis (List[List[int]]): list of basis states :math:`|b_i\rangle`
            wires (Sequence[int]): list of wires that the operator acts on
            work_wire (Union[Wires, int, str]): the auxiliary wire used for the permutation
```

File: tests/templates/test_state_preparations/test_superposition.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+        def circuit():
+            qml.Superposition(
+                np.sqrt(probs), basis=basis, wires=range(len(basis[0])), work_wire=len(basis[0])
+            )
+            return qml.probs(range(len(basis[0])))
+
+        output = circuit()
+        for i, base in enumerate(basis):
+            dec = int("".join(map(str, base)), 2)
+            assert np.isclose(output[dec], probs[i])
```
Comment (soranjh) [inline tests/templates/test_state_preparations/test_superposition.py:92]: You cannot just assert that the returned probs match? Why need to create the int?

**Review by soranjh — CHANGES_REQUESTED** [review]
Thanks @KetpuntoG, left some comments but looks good overall.

File: tests/templates/test_state_preparations/test_superposition.py (hunk 7 above)
Comment (KetpuntoG) [reply tests/templates/test_state_preparations/test_superposition.py:92]: The reason is that although the coefficients are the same, one is compressed and the other is in 2^n format with many zeros. Through the decimal representation we can check the position where we have to compare

File: pennylane/templates/state_preparations/superposition.py [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+        >>> print(circuit())
+        [0.33333333 0.         0.33333333 0.         0.         0.
+        0.         0.33333333]
+
+
+    .. details::
+        :title: Details
+
+        The construction of this template is divided into two blocks.
+        The first block takes the list of coefficients :math:`c_i` and prepares the state:
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:165]:
```suggestion
        The input superposition state , :math:`|\phi\rangle = \sum_i^m c_i |b_i\rangle`, is implemented in two steps. First, the coefficients :math:`c_i` are used to prepares the state:
```

File: pennylane/templates/state_preparations/superposition.py [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+
+        .. math::
+
+            |\phi\rangle = \sum_i^m c_i |i\rangle,
+
+        where :math:`|i\rangle` are the computational basis states and :math:`m` is the number of terms
+        in the superposition. This is done using the
+        :class:`~.StatePrep` template in the fisrt :math:`\lceil \log_2 m \rceil` qubits.
+        This significantly reduces complexity by not having to work on the entire system.
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:174]:
```suggestion
        where :math:`|i\rangle` is a computational basis states and :math:`m` is the number of terms
        in the superposition. This is done using the
        :class:`~.StatePrep` template in the fisrt :math:`\lceil \log_2 m \rceil` qubits. Note that the number of qubits depends on the number of terms in the superposition, which helps to reduce the complexity of the operation.
```

File: pennylane/templates/state_preparations/superposition.py [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+
+        The second block is responsible for the permutation of the basis states previously prepared to
+        the target basis states.
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:177]:
```suggestion
        The second step permutes the basis states prepared previously to
        the target basis states.
```

File: pennylane/templates/state_preparations/superposition.py [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+
+        .. math::
+
+            |i\rangle \rightarrow |b_i\rangle.
+
+        This block maps the elements one by one using an auxiliary qubit.
+        The process can be divided into three operations. Let's assume
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:184]:
```suggestion
        This process can be done in three separate operations. Let's assume
```

File: pennylane/templates/state_preparations/superposition.py [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+        we want to map :math:`|i\rangle` to :math:`|b_i\rangle`:
+
+        1. By using a multi-controlled NOT gate, we check if the input state is :math:`|i\rangle` and
+        store the information in the auxiliary qubit (if the state is :math:`|i\rangle` the auxiliary
+        qubit will be in the :math:`|1\rangle` state).
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:189]:
```suggestion
        1. By using a multi-controlled NOT gate, check if the input state is :math:`|i\rangle` and
        store the information in the auxiliary qubit. If the state is :math:`|i\rangle` the auxiliary
        qubit will be in the :math:`|1\rangle` state.
```

File: pennylane/templates/state_preparations/superposition.py [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+
+        2. If the auxiliary qubit is in the :math:`|1\rangle` state, we modify the input state by adding
+        PauliX gates in the bits that are different between :math:`|i\rangle` and :math:`|b_i\rangle`.
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:192]:
```suggestion
        2. If the auxiliary qubit is in the :math:`|1\rangle` state, the input state is modified by applying
        ``X`` gates to the bits that are different between :math:`|i\rangle` and :math:`|b_i\rangle`.
```

File: pennylane/templates/state_preparations/superposition.py [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+
+        3. By using a multi-controlled NOT gate, we check if the final state is :math:`|b_i\rangle` and
+        return to :math:`|0\rangle` the auxiliary qubit.
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:195]:
```suggestion
        3. By using a multi-controlled ``NOT`` gate, check if the final state is :math:`|b_i\rangle` and
        return the auxiliary qubit back to :math:`|0\rangle` state.
```

File: pennylane/templates/state_preparations/superposition.py [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+
+        Appliying these two blocks together results in the desired superposition:
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:197]:
```suggestion
        Applying all these together prepares the desired superposition:
```

File: pennylane/templates/state_preparations/superposition.py [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+
+        .. math::
+
+            |\phi\rangle = \sum_i^m c_i |b_i\rangle.
+
+        The decomposition has a complexity that grows linearly with the number of terms,
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:203]:
```suggestion
        The decomposition has a complexity that grows linearly with the number of terms in the superposition,
```

File: pennylane/templates/state_preparations/superposition.py [hunk 17]
```diff
@@ -0,0 +1,335 @@
...
+    ndim_params = (1,)
+
+    def __init__(
+        self, coeffs, basis, wires, work_wire, id=None
+    ):  # pylint: disable=too-many-positional-arguments, too-many-arguments
+
+        unique_basis = qml.math.unique(
+            qml.math.array([tuple(sub_array) for sub_array in basis]), axis=0
+        )
+
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:219]: Please put checks to make sure:
* the elements of the basis states cannot be any values other than 0 and 1
* the input superposition is normalised (if it has to be according to the decomposition)
* the basis states have the same length, so the user cannot put [[0, 1], [1, 0, 0]]

**Review by soranjh — CHANGES_REQUESTED** [review]
Thanks @KetpuntoG, left my final set of comments, will approve when they are addressed.

**Comment by KetpuntoG** [comment]
> Thanks @KetpuntoG, left my final set of comments, will approve when they are addressed.

Thanks for the review @soranjh , I worked on the changes and added the checks. Note that the normalization check, verify that we are not with an abstract input, otherwise it doesn't work with jit (same as other templates)

File: pennylane/templates/state_preparations/superposition.py [hunk 18]
```diff
@@ -0,0 +1,345 @@
...
+        |\phi\rangle = \sum_i^m c_i |b_i\rangle.
+
+    See the Details section for more information about the decomposition.
+
+    Args:
+        coeffs (List[float]): list of coefficients :math:`c_i` normalized
+        basis (List[List[int]]): list of basis states :math:`|b_i\rangle` with the same length
+        wires (Sequence[int]): list of wires that the operator acts on
+        work_wire (Union[Wires, int, str]): the auxiliary wire used for the permutation
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:137]:
```suggestion
        coeffs (List[float]): normalized coefficients of the superposition
        basis (List[List[int]]): basis states of the superposition
        wires (Sequence[int]): wires that the operator acts on
        work_wire (Union[Wires, int, str]): the auxiliary wire used for the permutation
```

File: pennylane/templates/state_preparations/superposition.py [hunk 19]
```diff
@@ -0,0 +1,345 @@
...
+
+    .. math::
+
+        |\phi\rangle = \sum_i^m c_i |b_i\rangle.
+
+    See the Details section for more information about the decomposition.
+
+    Args:
+        coeffs (List[float]): list of coefficients :math:`c_i` normalized
+        basis (List[List[int]]): list of basis states :math:`|b_i\rangle` with the same length
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:135]: We should use the plural term "bases" instead of singular "basis", same way that we use coeffs as plural.

File: pennylane/templates/state_preparations/superposition.py [hunk 20]
```diff
@@ -0,0 +1,345 @@
...
+        .. math::
+
+            |\phi\rangle = \sum_i^m c_i |i\rangle,
+
+        where :math:`|i\rangle` is a computational basis states and :math:`m` is the number of terms
+        in the superposition. This is done using the
+        :class:`~.StatePrep` template in the fisrt :math:`\lceil \log_2 m \rceil` qubits. Note that the number of qubits depends on the number of terms in the superposition, which helps to reduce the complexity of the operation.
+
+        The second step permutes the basis states prepared previously to
+        the target basis states.
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:175]:
```suggestion
        the target basis states:
```

File: pennylane/templates/state_preparations/superposition.py [hunk 21]
```diff
@@ -0,0 +1,345 @@
...
+
+        .. math::
+
+            |i\rangle \rightarrow |b_i\rangle.
+
+        This block maps the elements one by one using an auxiliary qubit.
+        This process can be done in three separate operations. Let's assume
+        we want to map :math:`|i\rangle` to :math:`|b_i\rangle`:
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:183]:
```suggestion
        This can be done in three separate steps:
```

File: pennylane/templates/state_preparations/superposition.py [hunk 22]
```diff
@@ -0,0 +1,345 @@
...
+        return the auxiliary qubit back to :math:`|0\rangle` state.
+
+        Applying all these together prepares the desired superposition:
+
+        .. math::
+
+            |\phi\rangle = \sum_i^m c_i |b_i\rangle.
+
+        The decomposition has a complexity that grows linearly with the number of terms in the superposition,
+        unlike other methods such as :class:`~.MottonenStatePreparation`, that grows exponentially
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:202]:
```suggestion
        unlike other methods such as :class:`~.MottonenStatePreparation` that grows exponentially
```

**Review by soranjh — APPROVED** [review]
Looks gear, thanks @KetpuntoG.

File: pennylane/templates/state_preparations/superposition.py [hunk 23]
```diff
@@ -0,0 +1,344 @@
...
+
+    .. math::
+
+        |\phi\rangle = \sum_i^m c_i |b_i\rangle.
+
+    See the Details section for more information about the decomposition.
+
+    Args:
+        coeffs (List[float]): normalized coefficients of the superposition
+        basis (List[List[int]]): basis states of the superposition
```
Comment (josh146) [inline pennylane/templates/state_preparations/superposition.py:135]: @KetpuntoG do these need to be lists, or can they be arrays? I'm just thinking of JAX support

Comment (josh146) [reply pennylane/templates/state_preparations/superposition.py:135]: Thanks Soran! We would likely also want to make sure that `(j)np.array([[0, 1, 1], [1, 1, 0], ...])` works, because:

- the user may be creating this data programmatically with numpy,
- when using JIT, this forces arguments to be jax arrays, and
- large arrays are more efficient than lists

Comment (soranjh) [reply pennylane/templates/state_preparations/superposition.py:135]: We just need to update the doc and the example. The following already works:
```
coeffs = jnp.sqrt(jnp.array([1/3, 1/3, 1/3]))
basis = jnp.array([[1, 1, 1], [0, 1, 0], [0, 0, 0]])
```

File: pennylane/templates/state_preparations/superposition.py [hunk 24]
```diff
@@ -0,0 +1,344 @@
...
+        bases (tensor-like[int]): basis states of the superposition
+        wires (Sequence[int]): wires that the operator acts on
+        work_wire (Union[Wires, int, str]): the auxiliary wire used for the permutation
+
+    **Example**
+
+    .. code-block::
+
+        coeffs = jnp.sqrt(jnp.array([1/3, 1/3, 1/3]))
+        bases = jnp.array([[1, 1, 1], [0, 1, 0], [0, 0, 0]])
```
Comment (soranjh) [inline pennylane/templates/state_preparations/superposition.py:144]:
```suggestion
        coeffs = np.sqrt(np.array([1/3, 1/3, 1/3]))
        bases = np.array([[1, 1, 1], [0, 1, 0], [0, 0, 0]])
```

**Review by soranjh — APPROVED** [review]
